lib/dry/schema/messages/abstract.rb

```diff
@@ -95,14 +92,24 @@ def rule(name, options = {})
         # @return [Template]
         #
         # @api public
-        def call(*args)
-          cache.fetch_or_store(args.hash) do
-            text, meta = lookup(*args)
+        def call(predicate, options)
+          cache.fetch_or_store(cache_key(predicate, options).hash) do
```

Comment:
Hi @solnic, this is still an incorrect fix to the issue raised by @aardvark179 in #132.

The problem is that 2 distinct Ruby objects (in this case arrays returned by the `cache_key` method) are allowed to have identical `hash` values.

The simplest fix for the issue is:
```diff
-cache.fetch_or_store(cache_key(predicate, options).hash) do
+cache.fetch_or_store(cache_key(predicate, options)) do
```

Are you worried that `predicate` and `options` wouldn't get GCed? Why are you using `hash` of the hash-key-object (instead of using the the hash-key-object itself as a key)?

---

```ruby
a = [1,2,3]
b = [2,3,4]

def a.hash; 138293478450784465; end # valid `hash` implementation, this can really happen
def b.hash; 138293478450784465; end # valid `hash` implementation, this can really happen

HASH_RIGHT = {}
HASH_RIGHT[a] = a
HASH_RIGHT[b] = b

HASH_WRONG = {}
HASH_WRONG[a.hash] = a
HASH_WRONG[b.hash] = b

puts HASH_RIGHT.inspect # {[1, 2, 3]=>[1, 2, 3], [2, 3, 4]=>[2, 3, 4]}
puts HASH_WRONG.inspect # {138293478450784465=>[2, 3, 4]}
```
---

Since some time now ruby randomizes its primitives' internal `hash` values by a private `seed` so that each new invocation of ruby VM produces different values for e.g. `puts 1.hash`. What this results in is that there is a __random probability__ for hash collision, any failures will not be reproducible and testing in general will not discover this issue.

Reply:
@thedarkone thanks, we've discovered that too and addressing it in #137